cognite/client/data_classes/data_modeling/shared.py

```diff
@@ -0,0 +1,199 @@
+            return ContainerReference.load(data)
+        elif data["type"] == "view":
+            return ViewReference.load(data)
+        raise ValueError(f"Type {data['type']} is not supported.")
+
+
+@dataclass
+class ContainerReference(Reference):
+    @classmethod
+    def load(cls, data: dict) -> ContainerReference:
+        return cast(ContainerReference, super()._load(data))
+
+    def dump(self, camel_case: bool = False, *_: Any, **__: Any) -> dict[str, str]:
+        output = super().dump(camel_case)
+        output["type"] = "container"
+        return output
+
+
+@dataclass
+class ViewReference(Reference):
+    version: str
+
+    @classmethod
+    def load(cls, data: dict) -> ViewReference:
+        return cast(ViewReference, super()._load(data))
+
+    def dump(self, camel_case: bool = False, *_: Any, **__: Any) -> dict[str, str]:
+        output = super().dump(camel_case)
+        output["type"] = "view"
+        return output
+
+
+@dataclass
+class PropertyType(ABC):
+    def dump(self, camel_case: bool = False, *_: Any, **__: Any) -> dict[str, str | dict]:
+        output = asdict(self)
+        return convert_all_keys_recursive(output, camel_case)
+
+    @classmethod
+    def _load(cls, data: dict) -> PropertyType:
+        return cls(**convert_all_keys_to_snake_case(data))
+
+    @classmethod
+    def load(
+        cls, data: dict, direct_type_cls: typing.Type[DirectNodeRelation] = None
+    ) -> Text | Primitive | CDFExternalIdReference | DirectNodeRelation:
+        if "type" not in data:
+            raise ValueError("Property types are required to have a type")
+
+        type_ = data["type"]
+
+        if type_ in PRIMITIVE_TYPE_SET:
+            return Primitive.load(data)
+        elif type_ in TEXT_TYPE_SET:
+            return Text.load(data)
+        elif type_ in CDF_TYPE_SET:
+            return CDFExternalIdReference.load(data)
+        elif type_ in DIRECT_TYPE:
+            if direct_type_cls is None:
+                raise NotImplementedError
+            return direct_type_cls.load(rename_and_exclude_keys(data, exclude={"type"}))
+        else:
+            raise ValueError(
+                f"Invalid type {type_}. Must be {PRIMITIVE_TYPE_SET | TEXT_TYPE_SET | CDF_TYPE_SET | DIRECT_TYPE}"
+            )
+
+
+@dataclass
+class ListablePropertyType(PropertyType):
+    is_list: bool = False
+
+
+@dataclass
+class Text(ListablePropertyType):
+    collation: str = "ucs_basic"
+
+    @classmethod
+    def load(cls, data: dict, *_: Any, **__: Any) -> Text:
+        return cast(Text, cls._load(rename_and_exclude_keys(data, aliases=_PROPERTY_ALIAS, exclude={"type"})))
+
+    def dump(self, camel_case: bool = False, *_: Any, **__: Any) -> dict[str, str | dict]:
+        output = rename_and_exclude_keys(super().dump(camel_case), aliases=_PROPERTY_ALIAS_INV)
+        output["type"] = "text"
+        return output
+
+
+@dataclass
+class Primitive(PropertyType):
+    type: PrimitiveType
```

Comment:
Why `Primitive` - why not just create classes for each type? "Text" is also a primitive type, so this doesn't make sense to me.